module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,369 @@
+            break;
+            case MIC_MSG_TEST_AUDIO:
+                char filename[1024];
+                sprintf(filename, "/home/nubots/nuhear/audio/voice_%03d.wav", msg.data);
+            
+                write_audio_to_file(handles.stdin, filename);
+            break;
+        }
+    });
+    
+    //THIS IS FOR TESTING
+    on<Trigger<SpeechIntentMessage>>().then([this](const SpeechIntentMessage &msg) {
+        //TODO needs to be freed...
+        printf("msg ptr = %p\n", (void*)&msg);
+        printf("text = %s (%p)\n", msg.text, msg.text);
+        printf("intent = %s (%p)\n", msg.intent, msg.intent);
+        for(Slot slot : msg.slots) {
+            printf("%s = %s\n", slot.name, slot.value);
+        }
+        //need to free the msg...
+        //delete &msg;
+    });
+
+    
+    on<IO>(handles.stdout, IO::READ).then([this] {
+        printf("fds[0].revents & POLLIN\n");
+            
+        //TODO: need to read and append to buffer until we reach a newline...
+        char buffer[0x1000];
+        ssize_t bytes_read = read(handles.stdout, buffer, sizeof(buffer)-1);
+        if(bytes_read > 0) {
+            assert((size_t)bytes_read < sizeof(buffer)); //TODO
+            buffer[bytes_read] = 0;
+                        
+            if(enabled) {            
+                char *intent_json = voice2json_recognize_intent(buffer);
+                if(!intent_json) {
+                    perror("error");
+                }
+                printf("%s\n", intent_json);
+                
+                std::unique_ptr<SpeechIntentMessage> intent = parse_voice2json_json(intent_json, strlen(intent_json));
+                emit(std::move(intent));
+
+                free(intent_json);
+            }
+            
+        }
+    });
+    
+    //TODO: have a better story on how we handle this
+    //for the moment we fatally error
+    on<IO>(handles.stderr, IO::READ).then([this] {
+       printf("fds[1].revents & POLLIN\n");
+            
+        char buffer[0x1000];
+        ssize_t bytes_read = read(handles.stderr, buffer, sizeof(buffer)-1);
+        if(bytes_read > 0) {
+            assert((size_t)bytes_read < sizeof(buffer)); //TODO
+            buffer[bytes_read] = 0;
+            
+            printf("STDERR (%ld) = %s\n", bytes_read, buffer);
+            
+            perror("stderr");
+        } 
+    });
+
+#if 1
+    //this is just for testing
+    //we need this as we do not have an
+    //environment for testing the microphone
+    on<IO>(STDIN_FILENO, IO::READ).then([this] {
+        char b = 0;
+        ssize_t bytes_read = read(STDIN_FILENO, &b, 1);
+        assert(bytes_read == 1);
+        switch(b) {
+            case 'q':
+                char *string = (char*)malloc(100);
+                size_t size = 100;
+                bytes_read = getline(&string, &size, stdin);
+                
+                uint32_t num = atoi(string);
+                printf("%ld,%d - %s \n", bytes_read, num, string);
```

Comment:
NUClear::log

Reply:
probably want to use `fmt` too:
`#include <fmt/format.h>`
then when you use it:
`[NUClear::]log(fmt::format("{},{} - {}", bytes_read, num, string));`

Reply:
Thanks, it's been changed to use NUClear::log and fmt::format now.